### PR TITLE
Add connection pool and database backed logging provider

### PR DESCRIPTION
**Pull request by mathiasgredal**
This PR refactors the DBConnection to use a database connection pool, which makes more efficient use of database resources since connections are reused. This PR also implements a LoggingProvider using Log4J and the new database pool, to make database backed logs.

### CODE REVIEW

File: src/main/java/dk/sdu/se_f22/sharedlibrary/db/LoggingProvider.java
Comment (MagnusHJensen) [inline src/main/java/dk/sdu/se_f22/sharedlibrary/db/LoggingProvider.java:25]: Is this going to be used? If not it should be removed, if it's going to be used it should be implemented.

File: src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBConnection.java
Comment (MagnusHJensen) [inline src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBConnection.java:72]: Maybe make the wrapper catch the exception? So users either get a Connection or null, so we don't have to put try-catch everywhere the db is accessed.

File: src/main/java/dk/sdu/se_f22/Main.java
Comment (MagnusHJensen) [inline src/main/java/dk/sdu/se_f22/Main.java:7]: Should all classes use this instance of the logger? Or should they create their own, with their respective class, that the logging comes from?

**Review by MagnusHJensen — CHANGES_REQUESTED** [review]

File: src/main/java/dk/sdu/se_f22/Main.java
Comment (mathiasgredal) [reply src/main/java/dk/sdu/se_f22/Main.java:7]: I think the convention in Log4J is that each class has its own static logger. https://logging.apache.org/log4j/2.x/manual/api.html

File: src/main/java/dk/sdu/se_f22/sharedlibrary/db/LoggingProvider.java
Comment (mathiasgredal) [reply src/main/java/dk/sdu/se_f22/sharedlibrary/db/LoggingProvider.java]: This should be fixed now.

**Comment by MagnusQuist** [comment]
What's the point of having a connection-pool when we're never going to have a shared database? (Which is really dumb)
Each individual is going to have a local database, so there won't ever be need for multiple db connections.

**Comment by mathiasgredal** [comment]
> What's the point of having a connection-pool when we're never going to have a shared database? (Which is really dumb) Each individual is going to have a local database, so there won't ever be need for multiple db connections.
> 
> I guess it makes sense if we're simulating users going through the whole sequence (all modules)

It is generally a bad practice to have a database connection open for the entire lifecycle of the program and if any groups decide to implement multithreading, they don't have to wait on the other threads to get a connection to the database. Also this library precaches the compiled SQL statements, which can give slight performance improvements for repeatedly run statements.

**Comment by MagnusQuist** [comment]
> > What's the point of having a connection-pool when we're never going to have a shared database? (Which is really dumb) Each individual is going to have a local database, so there won't ever be need for multiple db connections.
> > I guess it makes sense if we're simulating users going through the whole sequence (all modules)
> 
> It is generally a bad practice to have a database connection open for the entire lifecycle of the program and if any groups decide to implement multithreading, they don't have to wait on the other threads to get a connection to the database. Also this library precaches the compiled SQL statements, which can give slight performance improvements for repeatedly run statements.

Having a connection open for the entirety of the program can easily be corrected just by closing the connection after use. I get your point on threading being a possible problem, though i fail to see where groups would implement that. But true, we have to account for it anyway.

Slight performance improvements are not important for this program if they confuse the other students. Everyone should be able to understand the code in the project, especially the shared code which all groups are going to use.

**Comment by mathiasgredal** [comment]
> > > What's the point of having a connection-pool when we're never going to have a shared database? (Which is really dumb) Each individual is going to have a local database, so there won't ever be need for multiple db connections.
> > > I guess it makes sense if we're simulating users going through the whole sequence (all modules)
> > 
> > 
> > It is generally a bad practice to have a database connection open for the entire lifecycle of the program and if any groups decide to implement multithreading, they don't have to wait on the other threads to get a connection to the database. Also this library precaches the compiled SQL statements, which can give slight performance improvements for repeatedly run statements.
> 
> Having a connection open for the entirety of the program can easily be corrected just by closing the connection after use. I get your point on threading being a possible problem, though i fail to see where groups would implement that. But true, we have to account for it anyway.
> 
> Slight performance improvements are not important for this program if they confuse the other students. Everyone should be able to understand the code in the project, especially the shared code which all groups are going to use.
> 
> I suggested the use of connection pooling earlier in the project, but it was refused due to complexity

I fail to see how a connection pool is too complex, since all the pooling aspects are abstracted away through a library. Usage is very simple, since you just need to get a database connection inside a try-with-resources and make your queries there. If the general consensus is that we don't want a connection pool, then we shouldn't have one, but i don't really see any downsides to having one.

**Comment by MagnusQuist** [comment]
> > > > What's the point of having a connection-pool when we're never going to have a shared database? (Which is really dumb) Each individual is going to have a local database, so there won't ever be need for multiple db connections.
> > > > I guess it makes sense if we're simulating users going through the whole sequence (all modules)
> > > 
> > > 
> > > It is generally a bad practice to have a database connection open for the entire lifecycle of the program and if any groups decide to implement multithreading, they don't have to wait on the other threads to get a connection to the database. Also this library precaches the compiled SQL statements, which can give slight performance improvements for repeatedly run statements.
> > 
> > 
> > Having a connection open for the entirety of the program can easily be corrected just by closing the connection after use. I get your point on threading being a possible problem, though i fail to see where groups would implement that. But true, we have to account for it anyway.
> > Slight performance improvements are not important for this program if they confuse the other students. Everyone should be able to understand the code in the project, especially the shared code which all groups are going to use.
> > I suggested the use of connection pooling earlier in the project, but it was refused due to complexity
> 
> I fail to see how a connection pool is too complex, since all the pooling aspects are abstracted away through a library. Usage is very simple, since you just need to get a database connection inside a try-with-resources and make your queries there. If the general consensus is that we don't want a connection pool, then we shouldn't have one, but i don't really see any downsides to having one.

There are absolutly no downsides to having a connection pool. But having "all the pooling aspects are abstracted away through a library" results in the majority of people not understanding how it works.  I think we need to discuss this with other peers before implementing it. If we agree on implementing it, I'm all for it as it is the best approach.

**Review by MagnusHJensen — APPROVED** [review]
Just tagging on here.
The general PR is approved, it looks good.
But I don't know how to relate to the debate regarding to implement it or not.

**Comment by mathiasgredal** [comment]
I have made a clarifying comment on the `DBConnection` explaining how the class is implemented, what libraries it uses and what a connection pool is, along with links for further further research. I hope this makes it clearer for others why a connection pool was chosen and gives them a general idea on how it works. I do still think that it is still a good idea to get some input from the groups about the decision though.

**Comment by Autowinto** [comment]
I think it's a good idea having a connection pool. The understanding of how it works in practice, should come with the following Data Management lecture, so if a general lack of understanding is a concern, perhaps we should wait until after Monday to make the final decision?

**Comment by MagnusQuist** [comment]
> I think it's a good idea having a connection pool. The understanding of how it works in practice, should come with the following Data Management lecture, so if a general lack of understanding is a concern, perhaps we should wait until after Monday to make the final decision?

Are you sure we'll hear about connection pooling in the upcoming lecture? As far as I can see it'll only be about singular connections through java, as we've already implemented.

**Comment by The0mikkel** [comment]
A comment from our counselor (or what they are called)

If we can describe this update and the connection pool concept, in a couple of lines, then our counselor is totally fine, with us using this, without knowing all the small details, in a very common implementation.

So as long we can describe the overall concept and it's not pensum, then it's alright that we don't know everything about it. And that is in general. Of course, other counselors may have other opinions.
This applies especially when using off-the-shelf kinds of implementations. Such as, we don't need to know every working detail of Intellij - Example: We can describe how to compile code with IntelliJ, in an overall "Press this button, and the code is compiled..." and not how it is actually compiled to the minor detail.

So as with this, I'm alright with this update.
As far as I could understand, this does not require any current code to use the database, from each module, to be refactored?

**Comment by mathiasgredal** [comment]
> So as with this, I'm alright with this update. As far as I could understand, this does not require any current code to use the database, from each module, to be refactored?

Depends on the usecase. When using a connection pool you usually get a connection make your query and close the connection as soon as possible to return the connection to the connection pool. In practice this would be done using a try-with-resources.

If your module gets single connection at the beginning and uses that for the entire module throughout the lifetime of the program(ie. getting the connection in the constructor and setting it as a class attribute), then there is no difference between the existing and the new implementation. And it is probably fine not to close the connection.

If you instead get a connection inside a for loop or a commonly called method, then we would quickly run out of connections and that code would have to be refactored. 

I haven’t checked how much of our code uses the second approach. If it is too much to refactor here and now, we can use a hybrid strategy, meaning that we reserve one connection from the pool and return that in the existing getConnection method. We then make a new method called getPooledConnection, that gets a connection from the connection pool, and explicitly states that these connections have to be closed. 

In this way we keep backwards compatibility with the existing code and when we add features we use the new approach while refactoring the old code to eventually use the pooled method everywhere, which will enable multi threading.

**Comment by The0mikkel** [comment]
Could we maybe get some information in the readme about the connection pool?
I have added information regarding how to input the details of how to connect (the properties) but we are missing the concept of how to use this connection

**Comment by The0mikkel** [comment]
Could we maybe get some information in the readme about the connection pool?
I have added information regarding how to input the details of how to connect (the properties) but we are missing the concept of how to use this connection

**Pull request merged**
